### DISCUSSION TITLE
Custom made UNSET value for query params stopped working after upgrade to 0.114.0

### DISCUSSION THREAD

**Discussion by rkazinets** [fastapi/fastapi#12146] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI


class UnsetType():
    ...


UNSET = UnsetType()

app = FastAPI()


@app.get("/foo")
def get_foo(bar: str = Query(UNSET)):
    if isinstance(bar, UnsetType):
        bar = ""
    return Response(content=bar)



async def test_get_foo(client: AsyncClient) -> None:
    response = await client.get("/foo")

    assert response.status_code == 200
    assert response.text == ""
```

### Description

In version 0.112.1 the test just would pass and 200 OK would be returned by the endpoint but as of 0.113.0 version it started to return 422 and following error message:
```
{
  "detail": [
    {
      "type": "string_type",
      "loc": [
        "query",
        "bar"
      ],
      "msg": "Input should be a valid string",
      "input": {}
    }
  ]
}
```

### Operating System

macOS

### FastAPI Version

0.113.0

### Pydantic Version

2.8.2

### Python Version

3.12.5

**Comment by YuriiMotov** [comment]
The reason of validation error you start getting is that since FastAPI 0.112.4 default value is validated (it didn't work before: see [discussion](https://github.com/fastapi/fastapi/discussions/11068)).

Since the parameter's type is `str`, but default value's type is `UnsetType` it doesn't pass validation.

So, it worked before 0.112.4 because of the bug, but now it doesn't work and it's expected (normal) behavior.

Don't know how to solve this in your case..

**Comment by yakubka** [comment]
The custom UNSET sentinel pattern broke because FastAPI moved to Pydantic v2 for parameter validation. In Pydantic v1, a sentinel object passed as a default was preserved as-is. In Pydantic v2, the validation layer coerces values to the declared type, so a sentinel that is not `None` and is not the correct type fails validation or gets replaced.

The supported way to represent an absent query parameter in Pydantic v2 is `None` with an `Optional` annotation.

```python
from fastapi import Query
from typing import Annotated

@app.get("/items")
async def get_items(
    status: Annotated[str | None, Query()] = None,
):
    if status is None:
        return {"message": "no filter"}
    return {"status": status}
```

If you need three states (absent, empty string, value), use a different query parameter name for each case, or parse `request.query_params` directly to distinguish between a missing key and an empty string value. The `request.query_params` dict does not go through Pydantic validation, so you have full control.
